includes/class-llms-media-protector.php

```diff
@@ -0,0 +1,640 @@
+			$content_type = $media_file->post_mime_type;
+			if ( 0 === stripos( $content_type, 'image/' ) ) {
+				// @todo Find or create an image to denote unauthorized access to an image file.
+				$file_name = LLMS_PLUGIN_DIR . 'assets/images/circle-with-a-slash.png';
+			} else {
+				header( 'HTTP/1.1 403 Forbidden' );
+				exit();
+			}
+		}
+
+		/**
+		 * Determine how the media file should be served.
+		 *
+		 * @since [version]
+		 *
+		 * @param string $serve_method One of the {@see LLMS_Media_Protector::SERVE_REDIRECT LLMS_Media_Protector::SERVE_X} constants.
+		 * @param int    $media_id     The post ID of the media file.
+		 */
+		$serve_method = apply_filters( 'llms_media_serve_method', self::SERVE_SEND_FILE, $media_id );
+
+		switch ( $serve_method ) {
+			case self::SERVE_READ_FILE:
+				$this->send_headers( $file_name, $media_id );
+				$this->read_file( $file_name );
+				break;
+			case self::SERVE_SEND_FILE:
+				$this->send_file( $file_name, $media_id );
+				break;
+			case self::SERVE_REDIRECT:
+			default:
+				$this->send_redirect( $media_id );
+				break;
+		}
+	}
+
+	/**
+	 * Sets the upload subdirectory.
+	 *
+	 * @since [version]
+	 *
+	 * @param string $upload_subdirectory
+	 * @return $this
+	 */
+	public function set_upload_subdirectory( string $upload_subdirectory ) {
+
+		// Add leading slash.
+		if ( strpos( $upload_subdirectory, '/' ) !== 0 ) {
+			$upload_subdirectory = '/' . $upload_subdirectory;
+		}
+
+		// Strip trailing slash.
+		$upload_subdirectory = rtrim( $upload_subdirectory, '/' );
+
+		$this->upload_subdirectory = $upload_subdirectory;
+
+		return $this;
+	}
+
+	/**
+	 * Removes the authorization filter on the media file.
+	 *
+	 * @since [version]
+	 *
+	 * @param int    $media_id             The post ID of the media file.
+	 * @param string $authorization_filter The hook name of the filter that authorizes users to view media files.
+	 * @return bool True on success, false on failure.
+	 */
+	public function unprotect( int $media_id, string $authorization_filter ) {
+
+		return delete_post_meta( $media_id, self::AUTHORIZATION_FILTER_KEY, $authorization_filter );
+	}
+
+	/**
+	 * Filters the 'uploads' directory data.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $uploads {
+	 *     Array of information about the upload directory.
+	 *
+	 *     @type string       $path    Base directory and subdirectory or full path to upload directory.
+	 *     @type string       $url     Base URL and subdirectory or absolute URL to upload directory.
+	 *     @type string       $subdir  Subdirectory if uploads use year/month folders option is on.
+	 *     @type string       $basedir Path without subdirectory.
+	 *     @type string       $baseurl URL path without subdirectory.
+	 *     @type string|false $error   False or error message.
+	 * }
+	 * @return array
+	 */
+	public function upload_dir( array $uploads ) {
+
+		$uploads['path'] = $uploads['basedir'] . $this->upload_subdirectory . $uploads['subdir'];
```

Comment:
Just so I make sure I'm reading this right, the presence of `$uploads['subdir']` will maintain the year/month folder structure within the custom directory(s), right?

I think this makes sense but I want to make sure I'm reading it right.

Reply:
Yes. In my dev environment, uploaded files are saved to `wp-content/uploads/llms-uploads/social-learning/2021/11/`.